asapdiscovery-modeling/asapdiscovery/modeling/protein_prep_v2.py

```diff
@@ -0,0 +1,123 @@
+import abc
+from pathlib import Path
+from typing import Literal, Optional
+
+import yaml
+from asapdiscovery.data.openeye import oechem
+from asapdiscovery.data.utils import seqres_to_res_list
+from asapdiscovery.modeling.modeling import (
+    make_design_unit,
+    mutate_residues,
+    spruce_protein,
+    superpose_molecule,
+)
+from pydantic import BaseModel, Field, root_validator
+
+
+class ProteinPrepperBase(BaseModel):
+    """
+    Base class for protein preppers.
+    """
+
+    prepper_type: Literal["ProteinPrepperBase"] = Field(
+        "ProteinPrepperBase", description="The type of prepper to use"
+    )
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    @abc.abstractmethod
+    def _prep(self, *args, **kwargs) -> oechem.OEDesignUnit:
+        ...
+
+    def prep(self, *args, **kwargs) -> oechem.OEDesignUnit:
+        return self._prep(*args, **kwargs)
+
+    @abc.abstractmethod
+    def provenance(self) -> dict[str, str]:
+        ...
+
+
+class ProteinPrepper(ProteinPrepperBase):
+    """
+    Protein prepper class that uses OESpruce to prepare a protein for docking.
+    """
+
+    prepper_type: Literal["ProteinPrepper"] = Field(
+        "ProteinPrepper", description="The type of prepper to use"
+    )
+
+    align: Optional[oechem.OEMol] = Field(
+        None, description="Reference structure to align to."
+    )
+    ref_chain: Optional[str] = Field(None, description="Chain ID to align to.")
+    active_site_chain: Optional[str] = Field(None, description="Chain ID to align to.")
+    seqres_yaml: Optional[Path] = Field(
+        None, description="Path to seqres yaml to mutate to."
+    )
+    loop_db: Optional[Path] = Field(
+        None, description="Path to loop database to use for prepping"
+    )
+    oe_active_site_residue: Optional[str] = Field(
+        None, description="OE formatted string of active site residue to use"
+    )
+
+    @root_validator
+    @classmethod
+    def _check_align_and_chain_info(cls, values):
+        """
+        Check that align and chain info is provided correctly.
+        """
+        align = values.get("align")
+        ref_chain = values.get("ref_chain")
+        active_site_chain = values.get("active_site_chain")
+        if align and not ref_chain:
+            raise ValueError("Must provide ref_chain if align is provided")
+        if align and not active_site_chain:
+            raise ValueError("Must provide active_site_chain if align is provided")
```

Comment:
if a chain isn't provided but the structure only has one chain, we could infer it in theory. not sure if we want to have a method that tries to do this inference before failing the validator, or if we just put the onus fully on the user

Reply:
I think lets keep the guessing outside the class for modularity's sake,  that way however you want to interface with it you know you have to provide the chain, rather than the input in the right format to guess the chain.